src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -423,6 +423,8 @@ private ValueTask SendFrameLockAcquiredNonCancelableAsync(MessageOpcode opcode,
                 // the task, and we're done.
                 if (writeTask.IsCompleted)
                 {
+                    writeTask.GetAwaiter().GetResult();
+                    writeTask = new ValueTask(_stream.FlushAsync());
```

Comment:
We must wait until `FlushAsync` finishes as well before returning.

Should just mean checking `writeTask.IsCompleted` again.

Reply:
What if `writeTask.IsCompleted` is false in that case? Is ot ok that we return `WaitForWriteTaskAsync` with two flush tasks inside?

Reply:
I wasn't suggesting replacing the IsCompleted check on the flush.  Miha means this needs to be:
```C#
if (writeTask.IsCompleted)
{
    writeTask.GetAwaiter().GetResult();
    writeTask = new ValueTask(_stream.FlushAsync());
    if (writeTask.IsCompleted)
    {
        return writeTask;
    }
}
```
and I was suggesting we should actually do:
```C#
if (writeTask.IsCompleted)
{
    if (!writeTask.IsCompletedSuccessfully)
    {
        return writeTask;
    }
    writeTask.GetAwaiter().GetResult();

    writeTask = new ValueTask(_stream.FlushAsync());
    if (writeTask.IsCompleted)
    {
        return writeTask;
    }
}
```

Reply:
Katya asked about potential second flush inside `WaitForWriteTaskAsync` below; if we are not able to tell from the task itself, I guess we can e.g. pass a flag to `WaitForWriteTaskAsync` if we don't want to call flush again...

Reply:
Yes, as I noted earlier, the fix I'd proposed doesn't prevent a second flush in the case where the write completes synchronously and the flush completes asynchronously; it just makes it so that the operation is still self (previously you could end up with two concurrent flushes, which is very much not safe).  If you decide you care to avoid the second flush, then you'd need to either pass additional information into that shared method, or duplicate the method to have a separate one for when the write vs the flush is what's pending.

Reply:
Sounds good to me. Compiler is complaining about using writeTask twice though, so I reworked it a little.